### PR TITLE
Allow using Gradle profiler in cross version tests

### LINKED ISSUE #9997: Allow using gradle-profiler in our performance tests
We want to use gradle profiler to run our performance tests (or at least some of them). To this end, we would use a Java API to configure gradle-profiler.

---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by wolfs**
Fixes #9997

This PR requires https://github.com/gradle/gradle-profiler/pull/144 on the Gradle profiler side.

### CODE REVIEW

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts [hunk 1]
```diff
@@ -67,6 +67,9 @@ dependencies {
     implementation(testLibrary("jetty"))
     implementation(testFixtures(project(":core")))
     implementation(testFixtures(project(":toolingApi")))
+    implementation("org.gradle.profiler:gradle-profiler:1.0-SNAPSHOT") {
+        exclude(group = "com.google.guava", module = "guava")
```
Comment (paplorinc) [inline subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: how much effort would it be to unify the two versions?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+@Category(PerformanceRegressionTest)
+@CleanupTestDirectory
+class AbstractCrossVersionGradleProfilerPerformanceTest extends Specification {
+
+    private static def resultStore = new CrossVersionResultsStore()
+    private static def reporter = SlackReporter.wrap(resultStore)
+
+    @Rule
+    TestNameTestDirectoryProvider temporaryFolder = new PerformanceTestDirectoryProvider()
```
Comment (paplorinc) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: probably some of these can be final as well

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        runner.workingDir = temporaryFolder.testDirectory
+        runner.current = new UnderDevelopmentGradleDistribution(buildContext)
+        performanceTestIdProvider.testSpec = runner
+    }
+
+    CrossVersionPerformanceTestRunner getRunner() {
+        runner
+    }
+
+    static {
```
Comment (paplorinc) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: what's the reason for not including this in the `cleanup` method instead?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        InvocationSpec invocationSpec = experiment.getInvocation();
+        File workingDirectory = invocationSpec.getWorkingDirectory();
+        workingDirectory.mkdirs();
+        copyTemplateTo(experiment, workingDirectory);
+
+        if (experiment.getListener() != null) {
+            experiment.getListener().beforeExperiment(experiment, workingDirectory);
+        }
+
+        if (!(invocationSpec instanceof GradleInvocationSpec)) {
```
Comment (paplorinc) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: minor: this could probably go after the initialization of `invocationSpec`

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+// TODO:
+//   - use temporary folder for temp files in Gradle profiler (PidInstrumentation, )
+//   - remove usages of `deleteOnExit` from gradle-profiler
+public class GradleProfilerBuildExperimentRunner implements BuildExperimentRunner {
```
Comment (paplorinc) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: it would simplify review if the move and the change would be in separate commits

**Comment by wolfs** [comment]
This is ready enough now. I converted the up-to-date and some Android performance tests to use Gradle profiler in the background. Flamegraph and differential flame graph generation is working. The CSV and HTML reports are written by Gradle profiler and archived in Teamcity. Moreover, it is possible to specify some measured build operations for Gradle profiler, which are then shown in the reports.

Before we merge this, we need to release a non-snapshot version of Gradle profiler, so we can depend on it.

As next steps, I suggest
- Move the differential flame graph generation to Gradle profiler, so we don't need to do it in our performance infrastructure as all
- Allow using Async profiler instead of JFR for ad hoc runs
- Convert more of the performance tests to use Gradle profiler.

@lptr Can you take a look?

**Comment by wolfs** [comment]
Here is an ad-hoc performance run which shows the reports from Gradle profiler as well: https://builds.gradle.org/viewLog.html?buildId=25501438&buildTypeId=Gradle_Util_Performance_AdHocPerformanceScenarioLinux&tab=artifacts&branch_Gradle_Util_Performance=wolfs%2Fsnapshot-change-detection#%2Fresults%2Fperformance%2Fbuild%2Ftest-results-performanceAdhocTest.zip!%2Fflames%2F5.7-commit-34f67f7

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts (hunk 1 above)
Comment (lptr) [reply subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: We need to document the reasons here.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import org.gradle.performance.results.SlackReporter
+import org.gradle.test.fixtures.file.CleanupTestDirectory
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.junit.Rule
+import org.junit.experimental.categories.Category
+import spock.lang.Specification
+
+@Category(PerformanceRegressionTest)
+@CleanupTestDirectory
+class AbstractCrossVersionGradleProfilerPerformanceTest extends Specification {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: Why copy this instead of abstracting it into a shared type?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java [hunk 7]
```diff
@@ -16,216 +16,14 @@
...
-import org.gradle.internal.concurrent.CompositeStoppable;
-import org.gradle.performance.measure.MeasuredOperation;
 import org.gradle.performance.results.MeasuredOperationList;
-import org.gradle.util.GFileUtils;
 
-import java.io.File;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.concurrent.atomic.AtomicBoolean;
+public interface BuildExperimentRunner {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java]: Is this a general "build experiment" runner, or is this only used for performance tests? Looks like it's a generic thing but the only two implementations I see are for performance tests.

We should either call everything a build experiment or a performance test I think.

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts [hunk 8]
```diff
@@ -95,3 +95,7 @@ tasks.jar {
     from(files(deferred{ flamegraph.map { zipTree(it) } }))
 }
 
+configurations.all {
+    resolutionStrategy.cacheDynamicVersionsFor(1, TimeUnit.MINUTES)
+}
```
Comment (lptr) [inline subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: How does this help?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java [hunk 9]
```diff
@@ -95,6 +100,10 @@ public void run(BuildExperimentSpec experiment, MeasuredOperationList results) {
                     results.add(measuredOperation);
                 }
             });
+            if (jfrProfileTargetDir != null) {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: You already set the generator to `NOOP` if the target dir is `null` so there shouldn't be a need for this check.

It seems to me that `flameGraphGenerator` should know about the JFR target dir instead of keeping it here. Same for `Profiler`, which could be created in the constructor, too.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+import org.gradle.profiler.report.HtmlGenerator
+import org.gradle.test.fixtures.file.CleanupTestDirectory
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.junit.Rule
+import org.junit.experimental.categories.Category
+import spock.lang.Specification
+
+@Category(PerformanceRegressionTest)
+@CleanupTestDirectory
+class AbstractCrossVersionGradleProfilerPerformanceTest extends Specification {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: There should be some Javadoc on this and its sibling type to explain the difference (and the intention to use this one in future).

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+import org.gradle.internal.concurrent.CompositeStoppable;
+import org.gradle.performance.measure.MeasuredOperation;
+import org.gradle.performance.results.MeasuredOperationList;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PerformanceTestBuildExperimentRunner extends AbstractBuildExperimentRunner {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java]: I think it would make a lot of sense to add some documentation here and on the sibling class to explain why we have two implementations, and that this one is going away in favor of the new one.

We could consider adding a prefix here as well like `Legacy` or `GradleInternal` or whatever. Same for `AbstractCrossVersion[Legacy|GradleInternal]PerformanceTest`.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy [hunk 12]
```diff
@@ -62,10 +62,11 @@ class AbstractCrossVersionGradleProfilerPerformanceTest extends Specification {
             new CsvGenerator(new File(debugArtifactsDirectory, "benchmark.csv")),
             new HtmlGenerator(new File(debugArtifactsDirectory, "benchmark.html"))
         )
+        def slackReporter = reporter
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: What's the point of this?

File: .teamcity/common/performance-test-extensions.kt [hunk 13]
```diff
@@ -50,3 +50,9 @@ fun performanceTestCommandLine(task: String, baselines: String, extraParameters:
 fun distributedPerformanceTestParameters(workerId: String = "Gradle_Check_IndividualPerformanceScenarioWorkersLinux") = listOf(
         "-Porg.gradle.performance.buildTypeId=$workerId -Porg.gradle.performance.workerTestTaskName=fullPerformanceTest -Porg.gradle.performance.coordinatorBuildId=%teamcity.build.id% -PgithubToken=%github.ci.oauth.token%"
 )
+
+val individualPerformanceTestArtifactRules = """
+        subprojects/*/build/test-results-*.zip => results
+        subprojects/*/build/tmp/**/log.txt => failure-logs
+        subprojects/*/build/tmp/**/profile.log => failure-logs
```
Comment (lptr) [inline .teamcity/common/performance-test-extensions.kt]: Are we by any chance leaving behind `profile-out-7` etc. directories from previous runs that will be collected here? (Probably not, right?)

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleBuildExperimentSpec.groovy [hunk 14]
```diff
@@ -88,6 +91,11 @@ class GradleBuildExperimentSpec extends BuildExperimentSpec {
             this
         }
 
+        GradleBuilder measuredBuildOperations(List<String> measuredBuildOperations) {
+            this.measuredBuildOperations = ImmutableList.copyOf(measuredBuildOperations)
+            this
+        }
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleBuildExperimentSpec.groovy]: These will be ignored for "traditional" performance tests, right? We should probably fail if someone tried to set it in that case.

**Comment by wolfs** [comment]
There is also a performance coordinator run which has the test in them which use the Gradle profiler backend: https://builds.gradle.org/viewLog.html?buildId=25502080&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance&branch_Gradle_Check_Stage_ReadyforMerge=wolfs%2Fagp%2Fsanta-tracker-perf

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts (hunk 1 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: We don't need to exclude it any more, since we now use the Android version in Gradle profiler as well.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy (hunk 6 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: The little code which is shared didn't seem worth introducing the abstract superclass. I can introduce the class if you think it is worth it.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java (hunk 7 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java]: The current nomenclature in the performance project are:
- Scenario, e.g. `non-abi change on largeJavaMultiProject`
- Build Experiment, e.g. getting the measurements for Gradle 5.7 on a scenario
- Invocation, e.g running Gradle 5.7 on a scenario once

So `BuildExperimentRunner` is a perfectly good name in that context, since it runs one build experiment. Note that it does not run the whole performance scenario, nor conducts contains the test assertions. It also lives in the `org.gradle.performance` package, so I don't see a need to add `Performance` to it.

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts (hunk 8 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: I was hoping it helps while I am using snapshot version that CI picks up the newest published `1.0-SNAPSHOT` version, instead of whatever there is the local dependency cache. This needs to be remove once we published a non-changing Gradle profiler version (1.0?).

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy (hunk 12 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: If the subsequent `Closure` is invoked from the subclass it does not have access to the private `reporter` field and will fail at runtime. I could also make the field protected to work around the problem. What do you prefer?

File: .teamcity/common/performance-test-extensions.kt (hunk 13 above)
Comment (wolfs) [reply .teamcity/common/performance-test-extensions.kt]: No, we set the output directory explicitly: https://github.com/gradle/gradle/blob/2136f7d957f40e8546f64bb4940522c57d4d8b77/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java#L144-L146

The `tmp` directory should be cleaned up by `clean` as well, which we run on CI.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleBuildExperimentSpec.groovy (hunk 14 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleBuildExperimentSpec.groovy]: Currently, you can't set those operations on the runner itself, since there are two separate classes: `GradleProfilerCrossVersionPerformanceTestRunner` and `CrossVersionPerformanceTestRunner`.

`buildExperimentListeners` and `invocationCustomizers` can only be specified on `CrossVersionPerformanceTestRunner`, while `buildMutators` and `measuredBuildOperations` can only be specified on `GradleProfilerCrossVersionPerformanceTestRunner`.

I think that should be enough already.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java (hunk 9 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: Both things are now instantiated in the constructor. As soon as all of the flame graph handling lives in Gradle profiler itself, we don't need to do anything on the Gradle side anymore.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java (hunk 11 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java]: I added some documentation and rename the "old" classes to use the prefix `GradleInternal`.

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts (hunk 8 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: Removed.

File: subprojects/internal-performance-testing/internal-performance-testing.gradle.kts (hunk 1 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/internal-performance-testing.gradle.kts]: I removed the exclusion.

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy (hunk 6 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: Alright, I don't feel very strongly about this one.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java [hunk 15]
```diff
@@ -57,12 +58,16 @@
     private final String jfrProfileTargetDir;
     private final ProfilerFlameGraphGenerator flameGraphGenerator;
     private final BenchmarkResultCollector resultCollector;
+    private final Profiler profiler;
 
     public GradleProfilerBuildExperimentRunner(BenchmarkResultCollector resultCollector) {
         this.jfrProfileTargetDir = org.gradle.performance.fixture.Profiler.getJfrProfileTargetDir();
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: I think this can be a local variable now.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy (hunk 12 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/AbstractCrossVersionGradleProfilerPerformanceTest.groovy]: I think it's fine, was just surprised about it.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java (hunk 11 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestBuildExperimentRunner.java]: Thanks.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java (hunk 9 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: :+1:

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java (hunk 7 above)
Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/BuildExperimentRunner.java]: I don't know why but I apparently imagined that subclasses had `PerformanceTest` in their names, but obviously not. Sorry about that, please ignore the noise.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java (hunk 15 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/GradleProfilerBuildExperimentRunner.java]: Good catch, converted!
